janus_core/helpers/utils.py

```diff
@@ -229,3 +231,89 @@ def dict_remove_hyphens(dictionary: dict) -> dict:
         if isinstance(value, dict):
             dictionary[key] = dict_remove_hyphens(value)
     return {k.replace("-", "_"): v for k, v in dictionary.items()}
+
+
+def results_to_info(
+    struct: Atoms,
+    *,
+    properties: Collection[Properties] = (),
+    invalidate_calc: bool = False,
+) -> None:
+    """
+    Copy or move MLIP calculated results to Atoms.info dict.
+
+    Parameters
+    ----------
+    struct : Atoms
+        Atoms object to copy or move calculated results to info dict.
+    properties : Collection[Properties]
+        Properties to copy from results to info dict. Default is ().
+    invalidate_calc : bool
+        Whether to remove all calculator results after copying properties to info dict.
+        Default is False.
+    """
+    if not properties:
+        properties = get_args(Properties)
+
+    if struct.calc:
+        # Set default architecture from calculator name
+        arch = struct.calc.parameters["arch"]
+        struct.info["arch"] = arch
+
+        for key in properties & struct.calc.results.keys():
+            tag = f"{arch}_{key}"
+            value = struct.calc.results[key]
+            if key == "forces":
+                struct.arrays[tag] = value
+            else:
+                struct.info[tag] = value
+
+        # Remove all calculator results
+        if invalidate_calc:
+            struct.calc.results = {}
+
+
+def output_structs(
+    images: MaybeSequence[Atoms],
+    *,
+    set_info: bool = True,
+    write_results: bool = False,
+    properties: Collection[Properties] = (),
+    invalidate_calc: bool = False,
+    **kwargs,
+) -> None:
+    """
+    Copy or move calculated results to Atoms.info dict and/or write structures to file.
+
+    Parameters
+    ----------
+    images : MaybeSequence[Atoms]
+        Atoms object or a list of Atoms objects to interact with.
+    set_info : bool
+        True to set info dict from calculated results. Default is True.
+    write_results : bool
+        True to write out structure with results of calculations. Default is False.
+    properties : Collection[Properties]
+        Properties to copy from calculated results to info dict. Default is ().
+    invalidate_calc : bool
+        Whether to remove all calculator results after copying properties to info dict.
+        Default is False.
+    **kwargs
+        Keyword arguments passed to ase.io.write.
+    """
+    if isinstance(images, Atoms):
+        images = (images,)
+
+    if set_info:
+        for image in images:
+            results_to_info(
+                image, properties=properties, invalidate_calc=invalidate_calc
+            )
+    else:
+        # Label architecture even if not copying results to info
+        for image in images:
+            if image.calc:
+                image.info["arch"] = image.calc.parameters["arch"]
+
+    if write_results:
+        write(images=images, write_results=not invalidate_calc, **kwargs)
```

Comment:
Does `invalidate_calc` have a default? I would say it's confusing that this is explicitly `not invalidate_calc`, I can't work out why intuitively.

Reply:
`invalidate_calc` defaults to False now, meaning we keep the `calc.results` dictionary for each images normally.

If `invalidate_calc is True`, then there's no point setting `write(write_results=True)`, as there will be no results to write for any of the images.

You could argue for a separate option if `invalidate_calc is False`, which I did have originally.

However, one issue with this is that the options become confusing, due to the clashes with the `write_results` flag that triggers `write`, which actually made it impossible to pass `write_results` to `write` in some situations at the moment.

That's not a good enough reason on its own, but I then felt that the point of these changes more generally is to keep the Atoms object and written files consistent, so if we write out an image that had a results dict, it makes sense that we can read it in with the results dict, which requires `write(write_results=True)`.

Reply:
`not invalidate_calc` is now only a default, so it can be explicitly overwritten if the user wants (although if `invalidate_calc` is True, this won't do anything)